### PR TITLE
ArrayPrimitive rule: report variable/receiver types and factory methods

### LINKED ISSUE #1245: Rule: Prefer IntArray instead of Array<Int> to escape boxing
### Prerequisites
Since Kotlin has no notion of primitive types, there are speicalized Array instances for each Java primitive:

- `IntArray`
- `DoubleArray`
- `FloatArray`
- `ShortArray`
- `ByteArray`
- `LongArray`
- `CharArray`

_From this point I will refer to the arrays of primitive types as `Array<Primitive>`._

As stated here in [docs](https://kotlinlang.org/docs/reference/basic-types.html#arrays) and [stackoverflow](https://stackoverflow.com/questions/45090808/intarray-vs-arrayint-in-kotlin), using `Array<Primitive>` leads to implicit boxing and perfomance hit.

Thus, I think we should mark `Array<Primitive>` usages to help improve performance and educate non-Java newcomers

### Actual rule

Actual rule should detect these patterns:
- [x] Using `Array<Primitive>` as a parameter of the function
- [x] Using `Array<Primitive>` as an explicit return type
- [ ] Using `arrayOf()` factory with primitive types
- [ ] All references to `Array<Primitive>` types

These can be standalone rules or one rule tweakable from config

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #1245

### CODE REVIEW

File: detekt-rules-performance/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitiveSpec.kt [hunk 1]
```diff
@@ -88,4 +94,65 @@ class ArrayPrimitiveSpec : Spek({
             assertThat(subject.compileAndLint(code)).isEmpty()
         }
     }
+
+    describe("variable type") {
+        it("is Array<Primitive>") {
+            val code = "val foo: Array<Int>? = null"
+            assertThat(subject.compileAndLint(code)).hasSize(1)
```
Comment (cortinico) [inline detekt-rules-performance/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitiveSpec.kt:101]:
```suggestion
            assertThat(subject.compileAndLintWithContext(env, code)).hasSize(1)
```

File: detekt-rules-performance/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitiveSpec.kt [hunk 2]
```diff
@@ -88,4 +94,65 @@ class ArrayPrimitiveSpec : Spek({
...
+        }
+    }
+
+    describe("receiver type") {
+        it("is Array<Primitive>") {
+            val code = "fun Array<Long>.foo() { println(this) }"
+            assertThat(subject.compileAndLint(code)).hasSize(1)
```
Comment (cortinico) [inline detekt-rules-performance/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitiveSpec.kt:108]:
```suggestion
            assertThat(subject.compileAndLintWithContext(env, code)).hasSize(1)
```

File: detekt-rules-performance/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitiveSpec.kt [hunk 3]
```diff
@@ -88,4 +94,65 @@ class ArrayPrimitiveSpec : Spek({
...
+        }
+
+        it("is arrayOf(String)") {
+            val code = "fun foo(x: String) = arrayOf(x)"
+            assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
+        }
+
+        it("is intArrayOf()") {
+            val code = "fun test(x: Int) = intArrayOf(x)"
+            assertThat(subject.compileAndLint(code)).isEmpty()
```
Comment (cortinico) [inline detekt-rules-performance/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitiveSpec.kt:155]:
```suggestion
            assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
```

File: detekt-rules-performance/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitive.kt [hunk 4]
```diff
@@ -35,44 +41,45 @@ import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
  * @active since v1.2.0
  */
 class ArrayPrimitive(config: Config = Config.empty) : Rule(config) {
+    companion object {
+        private val primitiveTypes = PrimitiveType.NUMBER_TYPES.map { it.typeName.asString() }
+        private val arrayOfFunctionFqName = FqName("kotlin.arrayOf")
```
Comment (cortinico) [inline detekt-rules-performance/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitive.kt:46]: Follow-up for the future: the same behavior should probably be flagged also for `emptyArray()`

File: detekt-rules-performance/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitive.kt [hunk 5]
```diff
@@ -35,44 +41,45 @@ import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
  * @active since v1.2.0
```
Comment (cortinico) [inline detekt-rules-performance/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitive.kt:41]:
```suggestion
 * @requiresTypeResolution
 */
```

**Review by arturbosch — APPROVED** [review]

**Comment by cortinico** [comment]
Side note: this is a rule that transition from NOT requiring to requiring type resolution. This should probably be noted in the README @arturbosch ?

**Review by schalkms — APPROVED** [review]

File: detekt-rules-performance/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitive.kt [hunk 6]
```diff
@@ -32,47 +38,49 @@ import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
  * fun returningFunction(): DoubleArray { }
  * </compliant>
  *
- * @active since v1.2.0
```
Comment (BraisGabin) [inline detekt-rules-performance/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitive.kt:35]: We should not disable this rule.

Comment (cortinico) [reply detekt-rules-performance/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ArrayPrimitive.kt:35]: Yup, I picked the wrong line for the suggestion

**Review by arturbosch — APPROVED** [review]

**Pull request merged**